### ISSUE TITLE
Pipeline Parallelism Fails when stage input does not produce gradients w.r.t. to next stage

### ISSUE THREAD

**Issue by man2machine** [pytorch/torchtitan#2748]
I'm reposting this issue here, because it seems that torchtitan gets more visibility and is the source of much of the active development of distributed PyTorch.

Original Issue: https://github.com/pytorch/pytorch/issues/152827

TLDR: Pipeline parallelism fails if stage input does not have gradients produced

Consider the case where a outputs from each pipeline stage is passed to the next stage, but whether or not the output is used or not for a particular batch is conditional (based on the code of the model). Hence, in many cases (such as in conditional or mixture of expert models), these weights may not be used for a particular stage, thus resulting in an error from get_bwd_send_ops: "[{self.stage_index}] for chunk {bwd_chunk_id} has gradients {grad} and is expecting to send gradients to stage {grad_recv_stage}". Even if that input is later used by subsequent stages (more than 1 stage later) and produces gradients, it will still error out with pytorch pipeline parallelism as it is implemented right now.

As a result, such a model fails when using pipeline parallelism, although with FSDP (or no parallelism) it has no issues. This is caused because for each input tensor to a stage, after calculating the gradient for that tensor, if there is no gradient it produces that error, even if that tensor would otherwise be passed onto a subsequent stage that would result in gradients being produced.

Currently it uses the dist.isend to send the tensor, but in order to send None, a different asynchronous P2P commutation operation is needed, to be able to asynchronously send or recv objects (which may or may not be tensors).

It would be great if this can be implemented, as this pipeline parallelism is critical to achieving high throughput in distributed execution, and conditional or mixture of expert models are limited due to this bug. This also prevents metadata like routing indices, in MoE models from being passed from stage to stage as they are int and thus don't have gradients at all.

Versions

Pytorch 2.6+

**Comment by man2machine** [comment]
The fix for this I believe requires a change in the if statements inside the for loop that causes the error.

https://github.com/pytorch/pytorch/blob/e32efbd59171c46b2890a218f89fbe062f2cb040/torch/distributed/pipelining/stage.py#L488

Basically, if its a float and the grad is None, then set it to have a zero grad (which can be an option set in PipelineStage, similar to the PR submitted for this issue: https://github.com/pytorch/pytorch/issues/152827 and if it is an int or tensor that doesn't require grad, it just ignores the whole check and transfers the data as is. The building of the tensor metadata for stage to stage transfers, would require an additional field to denote if the tensor requires_grad or not, which would be based on the first meta forward pass in the shape inference code. Then if it does require_grad but doesn't have a grad, and the option to fill in unused parameters with dummy 0 gradients is set, then we do that, otherwise we throw an error. In the PR for the FSDP fix, there is an option model.set_reduce_scatter_unused_params(True) that implements the dummy 0 gradient fix for the equivalent error in FSDP that occurs.

Right now I work around this problem by adding in post forward hooks to fill in the dummy zero gradients manually if a parameter needs grad but is unused. Based on this, I can confirm that this fix would work in principle, it just needs to be integrated into the internal pytorch code.

This would solve many errors in cases where the gradient is not produced on the immediately subsequent stage, but instead on 2 or 3 stages later, which is something I commonly encountered in my research work.

**Comment by sanketpurandare** [comment]
Hi @man2machine, we are doing a major refactor of torch.pipelining. This is the [RFC](https://github.com/pytorch/pytorch/issues/172419) and this is the [PR stack](https://github.com/pytorch/pytorch/pull/177728). I believe it should fix this bug.

**Comment by man2machine** [comment]
Thanks for linking the RFC. Are you talking about 3.6? The error that I mentioned in this issue can happen regardless of whether or not DTensor is used.

The error in this issue has to do with requires_grad tensors that the PipelineStage initialization metadata tracks, that happen to not produce a gradient during a particular forward pass. In this case they should be filled in with dummy zero gradients.

For example consider the following situations:
 - Stage i produces gradients for some inputs but not others.
 - Stage i does not produce gradients w.r.t. its input but stage i + 1 does produce gradients w.r.t its inputs. During the backward pass stage i's gradients tensor needs to be filled in with a zeroes tensor otherwise stage i + 1's gradients will fail to flow past stage i all the way to the first stage.

Also this is unrelated, but if the Pipeline parallelism code is being refactored perhaps support for https://github.com/pytorch/pytorch/issues/159711 could be added. This issue is about supporting dicts of tensors instead of just tuples of tensors as outputs from a PipelineStage forward. The internal PipelineStage code already uses Pytorch's internal pytree utilities, but it does an unnecessary cast to tuple, which prevents dicts from being used.

**Comment by wwwjn** [comment]
Close this PR as we can centralize the discussion in pytorch main repo: https://github.com/pytorch/pytorch/issues/152827 . This issue is not torchtitan specific
